File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
    #include <print>
    #include <vector>

    class Solution {
    public:
        int static trap(std::vector<int>& height) {
            int maxArea = 0;
            int l=0; int r=height.size()-1;
            int lmax=0; int rmax =0;
            while (l<r)
            {
                if (height[l] < height[r])
                {
                    if (height[l] >= lmax)
                        lmax = height[l];
                    else
                        maxArea += lmax - height[l];
                    ++l;
                }
                else
                {
                    if (height[r] >= rmax)
                        rmax = height[r];
                    else
                        maxArea += rmax - height[r];
                    --r;
                }
            }
            return maxArea;

        }


    };

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        stack = []  # Stores indices
        max_area = 0
        
        # Append 0 to flush remaining bars out of the stack at the end
        for i, h in enumerate(heights + [0]):
            while stack and heights[stack[-1]] > h:
                height_idx = stack.pop()
                height = heights[height_idx]
                
                # Calculate width based on the remaining element on stack
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, height * width)
                
            stack.append(i)
            
        return max_area